File: src/server/credentials.ts
import { newId } from "./db";
import { keychain } from "./keychain";

/**
 * Credential references: SQLite rows only store an opaque ref string; this
 * module resolves refs to secret material held in Keychain. Secret shapes are
 * JSON blobs so each provider can evolve its own token set.
 */

export interface OAuthTokens {
  accessToken: string;
  refreshToken?: string;
  /** epoch milliseconds when accessToken expires */
  expiresAt?: number;
  /** non-secret extras, e.g. X user id, Meta user id */
  meta?: Record<string, string>;
}

export interface BasicCredentials {
  identifier: string;
  secret: string;
  /** non-secret extras, e.g. Bluesky PDS URL, DID */
  meta?: Record<string, string>;
  /** session tokens (Bluesky), refreshed lazily */
  accessJwt?: string;
  refreshJwt?: string;
}

export function makeCredentialRef(kind: "oauth" | "basic" | "r2", owner: string): string {
  return `${kind}.${owner}.${newId()}`;
}

export function storeSecret<T>(ref: string, value: T): void {
  keychain().set(ref, JSON.stringify(value));
}

export function readSecret<T>(ref: string | undefined | null): T | null {
  if (!ref) return null;
  const raw = keychain().get(ref);
  if (!raw) return null;
  try {
    return JSON.parse(raw) as T;
  } catch {
    return null;
  }
}

export function deleteSecret(ref: string | undefined | null): void {
  if (ref) keychain().remove(ref);
}

File: Blather/Security/Credentials.swift
import Foundation

struct OAuthTokens: Codable, Hashable, Sendable {
    var accessToken: String
    var refreshToken: String?
    var expiresAt: Double?
    var meta: [String: String]?
}

struct BasicCredentials: Codable, Hashable, Sendable {
    var identifier: String
    var secret: String
    var meta: [String: String]?
    var accessJwt: String?
    var refreshJwt: String?
}

enum Credentials {
    static func makeRef(kind: String, owner: String) -> String {
        "\(kind).\(owner).\(Time.newId())"
    }

    static func store<T: Encodable>(_ ref: String, value: T) throws {
        try Keychain.store.set(ref: ref, secret: JSONCodec.encode(value))
    }

    static func read<T: Decodable>(_ type: T.Type, ref: String?) -> T? {
        guard let ref, let raw = Keychain.store.get(ref: ref) else { return nil }
        return try? JSONCodec.decoder.decode(type, from: Data(raw.utf8))
    }

    static func delete(ref: String?) {
        guard let ref else { return }
        Keychain.store.remove(ref: ref)
    }
}
